doc/constraints.md

```diff
@@ -0,0 +1,88 @@
+## Video Constraints
+
+Video constraints are bi-directional messages that are exchanged between a given
+endpoint and the bridge to control the exchanged video quality. Note that the
+intention is to augment the format in the future with frame-rate and potentially
+a max resolution/frame-rate.
+
+###### Message format to the bridge
+
+An endpoint can request different qualities for different endpoints, so it sends
+a map of endpoint ids -> video constraints back to the bridge. In JSON notation
+the message can be represented like this:
+
+    [{
+      'id': endpoint-1,
+      'constraints': {
+        'idealHeight': resolution,
+        'preferredHeight': resolution,
+        'preferredFps': frame-rate
+      },
+      ...
+    }]
+
+The preferredHeight and preferredFps properties are additional parameters that
```

Comment:
I struggle to understand the semantics of `preferredHeight` and `preferredFps`. What does `idealHeight=720, preferredHeight=360` mean? Ideally I want 720, but I prefer 360?

Following the code I find that:
1. `prefferedHeight` is not set by the client using `ReceiverVideoConstraintsMessage` (which is in fact not handled at all?).
2. It is set to `VideoConstraintsCompatibility` to:
a) The value from the bridge config for selected endpoints when there's more than one selected endpoint
b) -1 for everything else

It is used for:
1. Ranking endpoints, with higher `preferredHeight` being ranked higher (it is essentially a boolean pushing the selected endpoints up the list (but only when more than one is selected))
2. Allocating bitrate where it sets the level of the first round of allocation (i.e. we first fill the `preferred` value for every endpoint before improving any above that).


I think this is a very confusing concept, and we should fix that while this is still only in the bridge. The compatibility layer attempts to hide the complexity of "stage view" vs "tile view", but instead of providing a simple concept ("these endpoints need to go in the top of the list because they are special (selected)") it sets this `preferredHeight` value that then gets interpreted as a boolean in the ranking.

I think we should:
1. Separate the use of `preferred` for ranking vs as a floor for the first stage of allocation.
2. Use meaningful names. 

Perhaps a model with "min", "max" and "selected" makes sense. I.e. the client provides a list of "selected" endpoints for which selection should be prioritized. For each endpoint it also gives a "max" resolution and a "min" resolution. The semantics of "max" is clear, while "min" means "satisfy the min of everyone before allocating above it".

All that being said, I don't expect any changes in this PR other than making the documentation accurate (`preferred` is very confusing, is not explained anywhere, and after carefully reading the code for a while I'm still not sure I understand it).

Reply:
I totally agree it's confusing and it therefore needs to be fixed while it's still contained in the bridge. Do you see this as part of this PR ?

I'd argue we should keep/have/continue this discussion in the video policy refactoring PR [here](https://github.com/jitsi/jitsi-videobridge/pull/1349) that I opened a few weeks ago to tackle exactly this. @bbaldino also has some interesting ideas how to model things and @JonathanLennox can probably chime in.

This PR only meant to provide "aggressive" layer suspension (i.e. disabling the video for participants that are outside of the last-n set). "Effective video constraints" is the means to achieve just that.

Reply:
> I totally agree it's confusing and it therefore needs to be fixed while it's still contained in the bridge. Do you see this as part of this PR ?

No, as I said I think we should just fix the documentation. The situation right now is this:
1. We deprecated the selected/pinned/maxResolution signaling, and we support them via a "compatibility" layer.
2. We introduced a new way to describe the constraints with the "ideal/preferred", which we both agree is confusing and shouldn't be used.
3. The new way is only partially supported by the bridge (it handles only "ideal").
4. The new way is not supported by the client.

This PR add documentation about the new way, describing parts of it that are not implemented anywhere. I think this only adds to the confusion.

Reply:
Thanks for explaining Boris, I've updated the documentation to reflect reality a bit better.